frame/staking-rewards/src/lib.rs

```diff
@@ -599,8 +620,119 @@ pub mod pallet {
 					.try_insert(*asset_id, inflation)
 					.map_err(|_| Error::<T>::ReductionConfigProblem)?;
 			}
+
 			Ok((rewards_btree_map, reductions))
 		}
+
+		// TODO: tests
+		pub(crate) fn reward_acumulation_hook_reward_update_calculation(
+			pool_id: T::RewardPoolId,
+			asset_id: T::AssetId,
+			reward: Reward<T::AssetId, T::Balance>,
+			now: u64,
+		) -> (T::AssetId, Reward<T::AssetId, T::Balance>) {
+			let elapsed_time = match now.safe_sub(&reward.last_updated_timestamp) {
+				Ok(elapsed_time) => elapsed_time,
+				Err(_) => {
+					Self::deposit_event(Event::<T>::RewardAccumulationError { pool_id, asset_id });
+					return (asset_id, reward)
+				},
+			};
+
+			// SAFETY(benluelo): RewardRate::period is non-zero.
+			let periods_surpassed = elapsed_time.div(&reward.reward_rate.period);
+
+			if periods_surpassed.is_zero() {
+				(asset_id, reward)
+			} else {
+				// NOTE(benluelo): used this instead of `safe_mul` since
+				// `T::Balance` can't be multiplied by a `u64` and the definition of
+				// `CheckedMul` constrains the `Rhs` type to `Self`, so `T::balance`
+				// can't `.safe_mul(u64)`
+				//
+				// this also allows us to accumulate up to the maximum amount before
+				// erroring, short-circuiting on any error found and returning the last known
+				// "good" value.
+				let new_total_rewards = iter::repeat(reward.reward_rate.amount)
+					// REVIEW(benluelo): this cast is *probably* safe, but
+					// should be verified somehow
+					.take(periods_surpassed as usize)
+					.try_fold(reward.total_rewards, |current_total_amount, reward_amount| {
+						match current_total_amount.safe_add(&reward_amount) {
```

Comment:
Looks like we are implementing multiplication in a complicated way, can we simplify?